src/core/lib/support/closure_ref.h

```diff
@@ -0,0 +1,248 @@
+    MemClosure<Scheduler, T, F>::Schedule, MemClosure<Scheduler, T, F>::Run};
+}
+
+template <class Scheduler, void (*F)(), typename Env = std::nullptr_t>
+ClosureRef<> MakeClosure(Env* env = nullptr) {
+  return ClosureRef<>(&impl::FnClosure<Scheduler, Env, F>::vtable, env);
+}
+
+template <class Scheduler, typename T, void (*F)(T),
+          typename Env = std::nullptr_t>
+ClosureRef<T> MakeClosure(Env* env = nullptr) {
+  return ClosureRef<int>(&impl::FnClosure1<Scheduler, Env, T, F>::vtable, env);
+}
+
+template <class Scheduler, typename T, void (T::*F)()>
+ClosureRef<> MakeClosure(T* p) {
+  return ClosureRef<>(&impl::MemClosure<Scheduler, T, F>::vtable, p);
+}
+
+//
+// SCHEDULERS
+//
+
+class AcquiresNoLocks {
+ public:
+  template <class T, class F>
+  static void Schedule(F&& f, T* env) {
+    f(env);
+  }
+  template <class T, class F>
+  static void Run(F&& f, T* env) {
+    f(env);
+  }
+};
+
+template <class F>
+void QueueOnExecCtx(F&& f);
+
+class RunInCurrentThread {
+ public:
+  template <class T, class F>
+  static void Schedule(F&& f, T* env) {
+    QueueOnExecCtx([f, env]() { f(env); });
+  }
+  template <class T, class F>
+  static void Run(F&& f, T* env) {
+    f(env);
+  }
+};
+
+class RunInCombiner {
+ public:
+  template <class T, class F>
+  static void Schedule(F&& f, T* env) {
+    env->combiner()->Schedule([f, env]() { f(env); });
+  }
+  template <class T, class F>
+  static void Run(F&& f, T* env) {
+    env->combiner()->Run([f, env]() { f(env); });
+  }
+};
+
+// Dummy combiner lock impl
+
+class Combiner {
+ public:
+  template <class F>
+  void Schedule(F&& f);
+  template <class F>
+  void Run(F&& f);
+};
+
+//
+// TEST CODE
+//
+
+void PrintLine();
+void PrintInt(int);
+
+class Foo {
+ public:
+  void Callback();
+
+  Combiner* combiner() { return &combiner_; }
+
+ private:
+  Combiner combiner_;
+};
+
+ClosureRef<> Hidden();
+
+void test() {
+  // simple closures around functions, member functions
+  ClosureRef<> print_line = MakeClosure<AcquiresNoLocks, PrintLine>();
+  ClosureRef<int> print_int = MakeClosure<AcquiresNoLocks, int, PrintInt>();
+  Foo foo;
+  ClosureRef<> foo_cb = MakeClosure<AcquiresNoLocks, Foo, &Foo::Callback>(&foo);
+
+  print_line.Run();
+  print_int.Run(42);
+  foo_cb.Run();
+
+  // exec context test
+  ClosureRef<> foo_cb_in_exec_ctx =
+      MakeClosure<RunInCurrentThread, Foo, &Foo::Callback>(&foo);
+  foo_cb_in_exec_ctx.Schedule();
+
+  // combiner lock test - picks up combiner from Foo
+  ClosureRef<> foo_cb_in_combiner =
+      MakeClosure<RunInCombiner, Foo, &Foo::Callback>(&foo);
+  foo_cb_in_combiner.Schedule();
+  // can pass in a raw function too, but need to provide an environment
+  // in this case something that provides a combiner() method
+  ClosureRef<> print_line_in_combiner =
+      MakeClosure<RunInCombiner, PrintLine>(&foo);
+  print_line_in_combiner.Schedule();
+
+  Hidden().Run();
+
+  // empty closure test
+  ClosureRef<> empty;
+  empty.Run();
```

Comment:
again, but_why?

Reply:
again, to check the generated code output: this inlines successfully to be just a call to abort()